tendermint/src/serializers.rs

```diff
@@ -144,3 +144,41 @@ where
     let v = Option::deserialize(deserializer)?;
     Ok(v.map(|Wrapper(a)| a))
 }
+
+/// Parse empty block id as None.
+pub(crate) fn parse_non_empty_block_id<'de, D>(
```

Comment:
I'm kind of hoping this will go away with #101. In which case maybe we'll want to use an explicit VoteValue enum with variants BlockId and Nil ...

Reply:
I actually don't think `enum` would help us in these cases, because it's adding a branch just like `Option`, you just move the branch from `Option` to your own type.
The core problem lies in the protobuf semantic of optional field and default values, and the inconsistency between json and binary version of protobuf encoding.
In the semantic of protobuf, basically, there's no separation between null value and default value, they can be encoded as the same(skip), but in the json version, the default value is encoded as is, not skipped or null, it is actually correct in protobuf, because a skipped or null value is the same as a default value, but when we use `Option` to distingish null value and default value differently, we need extra trouble.